File: ruby/lib/solana_mpp/intents/session_split.rb
# frozen_string_literal: true

module SolanaMpp
  module Intents
    class SessionSplit
      attr_reader :recipient, :bps

      def initialize(recipient:, bps:)
        SessionRequest.assert_required(recipient, 'split recipient')
        raise ArgumentError, 'split bps must be between 1 and 10000' unless bps.is_a?(Integer) && bps.positive? && bps <= 10_000

        @recipient = recipient
        @bps = bps
      end

      def to_h
        {
          'recipient' => recipient,
          'bps' => bps
        }
      end

      def self.from_h(value)
        raise ArgumentError, 'session split must be a Hash' unless value.is_a?(Hash)

        new(recipient: value.fetch('recipient', '').to_s, bps: value.fetch('bps', 0).to_i)
      end
    end
  end
end

File: php/src/Intent/SessionSplit.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Intent;

use InvalidArgumentException;

final class SessionSplit
{
    public function __construct(
        public readonly string $recipient,
        public readonly int $bps,
    ) {
        if ($recipient === '') {
            throw new InvalidArgumentException('split recipient is required');
        }
        if ($bps <= 0 || $bps > 10_000) {
            throw new InvalidArgumentException('split bps must be between 1 and 10000');
        }
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'recipient' => $this->recipient,
            'bps' => $this->bps,
        ];
    }
}
